### PR TITLE
[SPARK-35410][SQL] SubExpr elimination should not include redundant children exprs in conditional expression

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch fixes an issue when dealing with common expressions in conditional expressions such as `CaseWhen` during subexpression elimination.

For example, previously we find common expressions among conditions of `CaseWhen`, but children expressions are also counted into. We should not count these children expressions as common expressions.

### Why are the changes needed?

If the redundant children expressions are counted as common expressions too, they will be redundantly evaluated and miss the subexpression elimination opportunity.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @maropu @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:99]: Do we need to revise line 83 consistently?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:102]: Yea, revised the method comment. Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
BTW, `master` branch is currently broken at compilation.

**Comment by dongjoon-hyun** [comment]
`master` branch compilation is recovered. Could you rebase to the master branch, @viirya ?

**Comment by dongjoon-hyun** [comment]
Also it would be better to wait for Takeshi and Wenchen's review.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun! Yea, just rebased to the master branch. I will leave this open to wait for the review from @maropu and @cloud-fan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:88]: Just a question; even if we filter out the redundant expr (e.g., `(a + b)` in this case) here, the suboptimal (this PR pointed out) case still can happen if the expr, `(a + b)`, is added as a common one in the other part? I thought a query like this: ` Seq((1, 1, 1)).toDF("a", "b", "c").select(when($"a" + $"b" + $"c" > 0, $"a" + $"b" + $"c").when($"a" + $"b" + $"c" <= 0, $"a" + $"b"))`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:106]: Is this loop not expensive? It seems the time-complexity is big-O(`the total number of expr nodes in candidateExprs)` x `(candidateExprs.size)^2 `)?

**Review by maropu — COMMENTED** [review]
Is this a bug? It looks a kind of improvements to me.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: is this only a problem for conditional expression?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:106]: +1, but I don't have a better idea now...

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:106]: Yea, I considered this part but didn't come out better one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:88]: The so called common expressions must occur at all branches/values. So in the above case, `(a + b)` is actually the only one common expression among two values `$"a" + $"b" + $"c` and `$"a" + $"b"`.

**Comment by viirya** [comment]
> Is this a bug? It looks a kind of improvements to me.

You can consider it as an improvement, yea. Although from user perspective, it is somehow hard to distinguish them clearly.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: So far the only one I can think about.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:106]: BTW `candidateExprs` size should be a small number.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: Found a non-conditional example that still is an issue even with this update (a bit contrived, but I'm sure there's a real use case)

```
val myUdf = udf(() => {
  println("In UDF")
  1
}).withName("myUdf")
spark.range(1).withColumn("a", myUdf()).select(($"a" + $"a") / ($"a" + $"a")).show()
```
This generates subexpressions myUdf() and (myUdf() + myUdf()), even though only the second one is used.

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: Thanks @Kimahriman. I see. Let me also look at it. As it is non-conditional case, but looks like the similar case. Let me see if it can be solved similarly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:106]: Yea, okay. I don't have a idea, too... That's just a question.

**Comment by maropu** [comment]
> Is this a bug? It looks a kind of improvements to me.
>> You can consider it as an improvement, yea. Although from user perspective, it is somehow hard to distinguish them clearly.

I was just wondering if we need to backport this fix or not. I think the update of CSE-related code can affect the performance of user's queries easily (e.g., , performance penalties caused by the expensive loop), so IMO it's safe to merge it into master only.

**Comment by Kimahriman** [comment]
It is a bit of a performance regression in certain cases so that seems like a bug. We have heavily chained expressions in when clauses and I suspect (but haven't been able to prove yet because of the complexity) it's causing us some issues.

**Comment by Kimahriman** [comment]
I did actually hit a bug today where the when value was being evaluated even though the condition was false. I wasn't able to find the exact root cause yet but turning off subexpression elimination fixed the issue. It was basically `when(col.rlike(...), udf(col))`, but more complex on both sides so somehow the UDF was getting subexpression eval'd early and failed because it didn't match the regular expression

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: Oh, I figured out. This might be an issue since we have sub-expr elimination. We also need to remove redundant children exprs for non-conditional cases.

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: But the fix might be different. I will work on it locally and submit another fix for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:88]: Updated the comment.

**Comment by viirya** [comment]
> I did actually hit a bug today where the when value was being evaluated even though the condition was false. I wasn't able to find the exact root cause yet but turning off subexpression elimination fixed the issue. It was basically `when(col.rlike(...), udf(col))`, but more complex on both sides so somehow the UDF was getting subexpression eval'd early and failed because it didn't match the regular expression

I see. Normally, for catalyst expressions, it is performance issue only. But for UDF, we cannot expect the logic put in UDF by users. It is possibly the UDF fails unexpectedly in such cases. So looks like it is better to backport it as a bug fix.

**Comment by Kimahriman** [comment]
> > I did actually hit a bug today where the when value was being evaluated even though the condition was false. I wasn't able to find the exact root cause yet but turning off subexpression elimination fixed the issue. It was basically `when(col.rlike(...), udf(col))`, but more complex on both sides so somehow the UDF was getting subexpression eval'd early and failed because it didn't match the regular expression
> 
> I see. Normally, for catalyst expressions, it is performance issue only. But for UDF, we cannot expect the logic put in UDF by users. It is possibly the UDF fails unexpectedly in such cases. So looks like it is better to backport it as a bug fix.

I'll try to see if I can get a real breaking example

**Comment by Kimahriman** [comment]
I figured out how the UDF error is happening. Basically it's due to `val values = c.branches.map(_._2) ++ c.elseValue` in `commonChildrenToRecurse`.

I think it should actually be something like
```
val values = if (c.isDefined) {
  c.branches.map(_._2) ++ c.elseValue
} else {
  Nil
}
```
because if the `elseValue` isn't defined then you're never going to find a common value among all branches.

The problem is then that gets rid of the optimization of `when(length(col) > 0, col)`, because col was only being pulled out of the when value because of this "bug" I think. So I started thinking about this a little bit. Would it make sense to do something like track separately:
- Definitely evaluated expressions
- Optionally evaluated expressions

And then consider an expression to be pulled out as a subexpression if:
- The expression is definitely evaluated at least once AND
- The expression is definitely or optionally evaluated at least twice

Then `commonChildrenToRecurse` would add the common expressions to definitely evaluated, but you could add each individual expression to optionally evaluated so they would get deduped if possible too. Maybe a separate issue from this PR, but figured I'd bring it up

**Comment by viirya** [comment]
> I figured out how the UDF error is happening. Basically it's due to `val values = c.branches.map(_._2) ++ c.elseValue` in `commonChildrenToRecurse`.
> 
> I think it should actually be something like
> 
> ```
> val values = if (c.isDefined) {
>   c.branches.map(_._2) ++ c.elseValue
> } else {
>   Nil
> }
> ```
> 
> because if the `elseValue` isn't defined then you're never going to find a common value among all branches.

Good catch! Do you want to submit a separate PR for this fix?
 
> The problem is then that gets rid of the optimization of `when(length(col) > 0, col)`, because col was only being pulled out of the when value because of this "bug" I think. So I started thinking about this a little bit. Would it make sense to do something like track separately:
> 
>     * Definitely evaluated expressions
> 
>     * Optionally evaluated expressions
> 
> 
> And then consider an expression to be pulled out as a subexpression if:
> 
>     * The expression is definitely evaluated at least once AND
> 
>     * The expression is definitely or optionally evaluated at least twice
> 
> 
> Then `commonChildrenToRecurse` would add the common expressions to definitely evaluated, but you could add each individual expression to optionally evaluated so they would get deduped if possible too. Maybe a separate issue from this PR, but figured I'd bring it up

For optionally evaluated expressions, do you mean expressions only happen in a few conditions/values? Since it is not definitely evaluated, but as a subexpression it will be evaluated all the times. I think it is hard to make sure it is always good for performance.

**Comment by viirya** [comment]
> Yeah that was the idea was for things like conditions/values that are only sometimes evaluated, evaluate them as a subexpression IF that expression is used somewhere else that's always evaluated anyway. Right now (assuming the above patch is applied), extra conditions and values that are only sometimes evaluated might not be pulled out as subexpressions even if they could be. So you would never evaluate an expression eagerly if we weren't definitely going to evaluate it at some point. I can try to make a PR to explain what I mean (and fix the bug I mentioned)

Okay. Could you submit the bug fix as a separate PR? For the other idea, it is another improvement and it is better not to mix them together.

**Comment by viirya** [comment]
@maropu @cloud-fan Do you have other comments on this change? Thanks.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: Any more thoughts on this? Was the subexpr sorting supposed to address this?

Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:314]: It might need another fix. I'm working on it and will submit it after these PRs merged.

**Comment by viirya** [comment]
> Okay. Could you submit the bug fix as a separate PR? For the other idea, it is another improvement and it is better not to mix them together.

@Kimahriman Created a JIRA for the elseValue issue: https://issues.apache.org/jira/browse/SPARK-35449

**Comment by viirya** [comment]
Oh, BTW, I think SPARK-35449 is actually the bug you hit. This could be seen as an improvement as @maropu suggested.

**Comment by Kimahriman** [comment]
> Oh, BTW, I think [SPARK-35449](https://issues.apache.org/jira/browse/SPARK-35449) is actually the bug you hit. This could be seen as an improvement as @maropu suggested.

Yeah I think that's correct. Though I checked one of my queries and it generated 34 subexpressions and only used one of them. So depends if you consider that a bug or improvement hah

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (Kimahriman) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2872]: I think the fix for https://issues.apache.org/jira/browse/SPARK-35449 will break this, since it's really a "bug" that the case value is included in subexpression resolution without an else value. Not a huge deal, I can try to fix in my follow up once this is merged

**Comment by dongjoon-hyun** [comment]
For this one, we are going to revisit after https://github.com/apache/spark/pull/32586 ? Did I understand correctly, @viirya ?

**Comment by viirya** [comment]
> For this one, we are going to revisit after #32586 to be safe? Did I understand correctly, @viirya ?

I think they are orthogonal improvements and can be merged independently.

**Comment by dongjoon-hyun** [comment]
Both are addressing corner cases for `SubExpr`s. I mean they are touching the same problem domains.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (Kimahriman) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: One potentially unrelated thing I just noticed, do we need to keep track of all of the `Expressions` here as well (as in an `Expr -> Seq[Expression]` map)? It's really basically keeping the first `Expression` found, but the codegen looks like it uses the `Expression` hash (versus the semantic hash) to lookup subexpressions.

**Comment by viirya** [comment]
> Both are addressing corner cases for `SubExpr`s. I mean they are touching the same problem domains.

Yea, sure. I agree.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: You mean `equivalenceMap`?

Comment (Kimahriman) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: I don't mean add it directly to that here. I'm just thinking of a really stupid example, `when((col + 1) > 0, col + 1).otherwise(1 + col)`. Wouldn't `col + 1` and `1 + col` resolve as a common expression because they're semantically equal, but only `col + 1` is added to `equivalenceMap`, so during codegen `1 + col` wouldn't be resolved to the subexpression?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: `col + 1` and `1 + col` will both be recognized as subexpression.

Comment (Kimahriman) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: Yeah but won't the codgen stage not replace `1 + col` since only `col + 1` will be added to the `equivalenceMap` entry for `Expr(col + 1)`? For non commonExprs cases, both would be in `equivalenceMap` so that the codegen stage maps both of those expressions to the resulting subexpression. Again, not super related to this PR, but was the easiest place to ask

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:97]: Both `1 + col` and `col + 1` will be replaced with the extracted subexpression during codege. We don't just look of key at `equivalenceMap` when replacing with subexpression.

**Comment by viirya** [comment]
#32586 was merged. Can we look at this if it is good to go? Thanks. cc @cloud-fan @dongjoon-hyun @maropu

**Review by maropu — APPROVED** [review]
okay, this improvement looks fine to me.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Could you make a backport to branch-3.1, @viirya ? There was a conflict on it.

**Comment by viirya** [comment]
Sure. Thanks @dongjoon-hyun @maropu @cloud-fan @Kimahriman

**Comment by viirya** [comment]
Ah, as this could be considered as an improvement (https://github.com/apache/spark/pull/32559#pullrequestreview-660631482, https://github.com/apache/spark/pull/32559#issuecomment-844470762, ), we can just have it merged to master only.

**Comment by dongjoon-hyun** [comment]
Got it!

**Pull request closed without merging**
